### PR TITLE
Add functionality needed by Cluster Autoscaler to Kubemark Provider.

### PR DESCRIPTION
**Pull request by bskiba**
Make adding nodes asynchronous. Add method for getting target
size of node group. Add method for getting node group for node.
Factor out some common code.

**Release note**:
```
NONE
```

### CODE REVIEW

File: pkg/kubemark/controller.go [hunk 1]
```diff
@@ -207,35 +242,31 @@ func (kubemarkController *KubemarkController) addNodeToNodeGroup(nodeGroup strin
 }
 
 func (kubemarkController *KubemarkController) removeNodeFromNodeGroup(nodeGroup string, node string) error {
-	pods, err := kubemarkController.externalCluster.podLister.List(labels.Everything())
-	if err != nil {
-		return err
+	pod := kubemarkController.getPodByName(node)
+	if pod == nil {
+		return fmt.Errorf("can't delete node %s from nodegroup %s. Node does not exist", node, nodeGroup)
```
Comment (aleksandra-malinowska) [inline pkg/kubemark/controller.go]: As discussed offline, can we print a warning in this case and return nil? Returning an error results in stopping resize and failing a test if even one node was scaled down by Cluster Autoscaler in the meantime.

File: pkg/kubemark/controller.go [hunk 2]
```diff
@@ -48,10 +48,13 @@ const (
...
-	nodeTemplate    *apiv1.ReplicationController
-	externalCluster externalCluster
-	kubemarkCluster kubemarkCluster
-	rand            *rand.Rand
+	nodeTemplate           *apiv1.ReplicationController
+	externalCluster        externalCluster
+	kubemarkCluster        kubemarkCluster
+	rand                   *rand.Rand
+	createNodeQueue        chan string
+	nodeGroupQueueSizeLock sync.Mutex
```
Comment (shyamjvs) [inline pkg/kubemark/controller.go]: nit: Could you place this lock below the variable it's locking?

File: pkg/kubemark/controller.go [hunk 3]
```diff
@@ -98,7 +101,10 @@ func NewKubemarkController(externalClient kubeclient.Interface, externalInformer
 			nodesToDelete:     make(map[string]bool),
 			nodesToDeleteLock: sync.Mutex{},
 		},
-		rand: rand.New(rand.NewSource(time.Now().UTC().UnixNano())),
+		rand:                   rand.New(rand.NewSource(time.Now().UTC().UnixNano())),
+		createNodeQueue:        make(chan string, 1000),
+		nodeGroupQueueSizeLock: sync.Mutex{},
```
Comment (shyamjvs) [inline pkg/kubemark/controller.go]: Here too

File: pkg/kubemark/controller.go [hunk 4]
```diff
@@ -177,15 +197,30 @@ func (kubemarkController *KubemarkController) SetNodeGroupSize(nodeGroup string,
 		}
 	case delta > 0:
 		for i := 0; i < delta; i++ {
-			if err := kubemarkController.addNodeToNodeGroup(nodeGroup); err != nil {
-				return err
-			}
+			kubemarkController.nodeGroupQueueSizeLock.Lock()
```
Comment (shyamjvs) [inline pkg/kubemark/controller.go]: Any reason why you're not locking/unlocking once for the entire for loop?

**Comment by shyamjvs** [comment]
Few nits and lgtm.
One question - You seem to create the hollow-node RCs asynchronously (which is fine - as we discussed earlier), but why are you deleting them synchronously?

**Comment by bskiba** [comment]
Fixed the nits. 
Regarding synch deletes - this is mostly for simplicity as it wasn't needed so much for CA at the moment since in most cases we delete only one node at a time.
@aleksandra-malinowska @MaciekPytel

**Comment by shyamjvs** [comment]
That's fine, thanks.

/lgtm
/approve no-issue

**Pull request merged**